freqtrade/exchange/exchange.py

```diff
@@ -2149,6 +2235,66 @@ def dry_run_liquidation_price(
             raise OperationalException(
                 "Freqtrade only supports isolated futures for leverage trading")
 
+    @retrier
+    def get_leverage_tiers_for_pair(self, pair: str):
+        # When exchanges can load all their leverage tiers at once in the constructor
+        # then this method does nothing, it should only be implemented when the leverage
+        # tiers requires per symbol fetching to avoid excess api calls
+        if (
+            self._api.has['fetchLeverageTiers'] and
+            not self._ft_has['can_fetch_multiple_tiers'] and
```

Comment:
Regarding [this comment](https://github.com/freqtrade/freqtrade/pull/6373#discussion_r801281879), whether we access the api or not will depend on whether we can fetch all the tiers for all markets at once or not, some exchanges(Okex) can only fetch them one market at a time, we also only fetch the tier if we have not fetched it before, which means I'll probably have to add a flag for `false` in the leverage_tiers dictionary if we can't get it. 

Binance can fetch them all at once, so it will just execute the `else` branch in this method

Reply:
That'd mean we can't support okex futures, as we have to disable backtesting for exchanges that use "online" calls.

Reply:
> That'd mean we can't support okex futures (nor any other exchange that has this requirement), as we have to disable backtesting for exchanges that use "online" calls.

It's a private endpoint for binance, so we store the data in `binance_leverage_tiers.json`, we could do the same for Okex, 

> I'd instead have a "fetch_for_pairs" (or whatever) method we can call with the whitelist of backtesting, which then populates the fields correctly (obviously it'll be slow - but only at the start).

Yeah I'd like to make the functionality a bit better, so that there's less of a delay when you actually need to make a trade, I'm not sure if that needs to be done in this PR, it's more of an enhanced functionality type thing, because we can just load the tiers the same way we do with binance

> also - is this data for okex changing that often? maybe we can cache it the same way we do the binance data?

It already is, the methods that call this method only do so if the pair isn't already saved in `_leverage_tiers`

Reply:
all of these "static" data endpoints are supposed to be used by backtesting as well (even though that may not be implemented yet for all instances).

This means - we must be certain which methods are going to call the exchange and which ones aren't.

The way `get_max_pair_stake_amount()` for okex is implemented - it's potentially (almost certainly on the first call) going to call the exchange - which will simply result in an error when running hyperopt.

The same problem (but worse) happens in `get_max_leverage()` - where it can happen on every exchange.

Fetching things like max/min amounts and leverage must be offline calls using data that's available (ideally from markets - or cached elsewhere), otherwise they cannot be used in backtesting - which means backtesting would not work.

For okex - maybe a "call for required pairs" (like - `load_leveraged_tiers_for_pairs()`) can be added which is called for backtesting so we can fetch the data as required.

i don't really have an answer for the remainder though, as you MUST be careful where you add online calls to not break backtesting support (which you clearly didn't care in this PR so far).

Reply:
The okex leverage tiers are loaded the same way as binance now, it would probably be a better idea to load all the ones you need at the beginning though to get up to date information (maybe give an option for binance too? To load all the leverage tiers via the private endpoint, and default to the json file if the private endpoint won't work?). I don't know if that needs to be done immediately though

----------------------

For reference, I just wrote a script that called `fetchLeverageTiers` for every linear swap symbol, I split the list of symbols in 2 and ran it twice, so we could run it for most pairlists from the beginning

```
import ccxt
import json


okex = ccxt.okex()
okex.loadMarkets()

markets = okex.markets
swaps = []
for symbol, market in markets.items():
    if (market["swap"] and market["linear"]):
        swaps.append(market["symbol"])

swaps = swaps[:len(swaps)//2]
# swaps = swaps[len(swaps)//2:]

tiers = {}
for symbol in swaps:
    res = okex.fetchLeverageTiers(symbol)
    tiers[symbol] = res[symbol]

filename = './okex_leverage_tiers.json'

with open(filename, 'w+') as file:
    file.write(json.dumps(tiers))
```